pennylane/ops/qutrit/non_parametric_ops.py

```diff
@@ -565,19 +531,15 @@ def compute_matrix(subspace=None):  # pylint: disable=arguments-differ
 
         return mat / np.sqrt(2)
 
-    def adjoint(self):
-        op = THadamard(wires=self.wires, subspace=self.subspace)
+    @property
+    def has_adjoint(self):  # pylint: disable=arguments-renamed, invalid-overridden-method
+        return self.subspace is not None
 
+    def adjoint(self):
         if self.subspace is None:
-            op.inverse = not self.inverse
-
-        return op
+            raise qml.operation.AdjointUndefinedError
```

Comment:
You can think of `adjoint` and `pow` as being part of the decomposition flow.  If we have an adjoint, what do we decompose it to? If we have a power, what do we decompose it to?

So we always want `adjoint` and `pow` to return things that are more likely to be supported. We need a direction to prevent circular loops in the decomposition graph.

So we could *choose* to make adjoint more "fundamental" than power and allow power operations to decompose to adjoint, or we could choose to make power operations more fundamental and decompose adjoint operations to powers. But it's very important we don't do both.

Adjoints are more efficient, but powers are easy to combine. So far we have kept them separate, but I'm considering finally allowing powers to decompose to adjoints.

Reply:
> So we could choose to make adjoint more "fundamental" than power and allow power operations to decompose to adjoint, or we could choose to make power operations more fundamental and decompose adjoint operations to powers. But it's very important we don't do both.

Oh I see, thanks for the additional context! For most of the basis gates I can think of, their adjoints can be expressed as some power of them (even if that power is 1 or -1), so expressing everything as a power makes a lot of sense.

Reply:
I think choosing between power and adjoint as the more "fundamental" will take more discussion and thought. For the near future, we should leave it as if. If a user requests the power, then they want the power. If a user requests the adjoint, then they want the adjoint. We don't mix the two for them.

In a later PR, we can start to mix the two, but we would want to do so across PennyLane consistently.